### PR TITLE
Ref #7928. New external transports

### PR DESCRIPTION
**Pull request by pablogs9**
This PR proposes a new and simplified external transport approach based on 3 flags (UDP, TCP and serial) and the copy of external transports in the proper folder. Merge with:

- [ ] https://github.com/micro-ROS/micro-ros-build/pull/84

### CODE REVIEW

File: src/c/profile/transport/ip/udp/udp_transport_external.c.template [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+size_t uxr_write_udp_data_platform(
+        uxrUDPPlatform* platform,
+        const uint8_t* buf,
+        size_t len,
+        uint8_t* errcode)
+{
+    // Place here your write bytes platform code
+    // Return number of wrote bytes
```
Comment (julibert) [inline src/c/profile/transport/ip/udp/udp_transport_external.c.template]:
```suggestion
    // Return number of bytes written
```

File: src/c/profile/transport/ip/udp/udp_transport_external.c.template [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+size_t uxr_read_udp_data_platform(
+        uxrUDPPlatform* platform,
+        uint8_t* buf,
+        size_t len,
+        int timeout,
+        uint8_t* errcode)
+{
+    // Place here your read bytes platform code
+    // Return number of read bytes (max bytes: len)
```
Comment (julibert) [inline src/c/profile/transport/ip/udp/udp_transport_external.c.template]:
```suggestion
    // Return number of bytes read (max bytes: len)
```

File: src/c/profile/transport/serial/serial_transport_external.c.template [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+bool uxr_close_serial_platform(struct uxrSerialPlatform* platform)
+{
+    // Place here your close platform code
+    // Return true if success
+}
+
+size_t uxr_write_serial_data_platform(uxrSerialPlatform* platform, uint8_t* buf, size_t len, uint8_t* errcode)
+{
+    // Place here your write bytes platform code
+    // Return number of wrote bytes
```
Comment (julibert) [inline src/c/profile/transport/serial/serial_transport_external.c.template]:
```suggestion
    // Return number of bytes written
```

File: src/c/profile/transport/serial/serial_transport_external.c.template [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+size_t uxr_read_serial_data_platform(uxrSerialPlatform* platform, uint8_t* buf, size_t len, int timeout, uint8_t* errcode)
+{
+    // Place here your read bytes platform code
+    // Return number of read bytes (max bytes: len)
```
Comment (julibert) [inline src/c/profile/transport/serial/serial_transport_external.c.template]:
```suggestion
    // Return number of bytes read (max bytes: len)
```

File: src/c/profile/transport/ip/tcp/tcp_transport_external.c.template [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        const char* port)
+{
+    // Place here your initialization platform code
+    // Return true if success
+}
+
+bool uxr_close_tcp_platform(
+        struct uxrTCPPlatform* platform)
+{
+    // Place here your close platform code
```
Comment (julibert) [inline src/c/profile/transport/ip/tcp/tcp_transport_external.c.template]:
```suggestion
    // Place here your closing platform code
```

File: src/c/profile/transport/ip/tcp/tcp_transport_external.c.template [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    // Return true if success
+}
+
+size_t uxr_write_tcp_data_platform(
+        struct uxrTCPPlatform* platform,
+        const uint8_t* buf,
+        size_t len,
+        uint8_t* errcode)
+{
+    // Place here your write bytes platform code
```
Comment (julibert) [inline src/c/profile/transport/ip/tcp/tcp_transport_external.c.template]:
```suggestion
    // Place here your writing bytes platform code
```

File: src/c/profile/transport/ip/tcp/tcp_transport_external.c.template [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+size_t uxr_read_tcp_data_platform(
+        struct uxrTCPPlatform* platform,
+        uint8_t* buf,
+        size_t len,
+        int timeout,
+        uint8_t* errcode)
+{
+    // Place here your read bytes platform code
```
Comment (julibert) [inline src/c/profile/transport/ip/tcp/tcp_transport_external.c.template]:
```suggestion
    // Place here your reading bytes platform code
```

File: src/c/profile/transport/ip/tcp/tcp_transport_external.c.template [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    // Return number of read bytes (max bytes: len)
```
Comment (julibert) [inline src/c/profile/transport/ip/tcp/tcp_transport_external.c.template]:
```suggestion
    // Return number of bytes read (max bytes: len)
```

File: src/c/profile/transport/ip/tcp/tcp_transport_external.c.template [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+size_t uxr_write_tcp_data_platform(
+        struct uxrTCPPlatform* platform,
+        const uint8_t* buf,
+        size_t len,
+        uint8_t* errcode)
+{
+    // Place here your write bytes platform code
+    // Return number of wrote bytes
```
Comment (julibert) [inline src/c/profile/transport/ip/tcp/tcp_transport_external.c.template]:
```suggestion
    // Return number of bytes written
```

File: src/c/profile/transport/ip/tcp/tcp_transport_external.c.template [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        uint8_t* errcode)
+{
+    // Place here your read bytes platform code
+    // Return number of read bytes (max bytes: len)
+}
+
+void uxr_disconnect_tcp_platform(
+        struct uxrTCPPlatform* platform)
+{
+    // Place here your disconnect platform code
```
Comment (julibert) [inline src/c/profile/transport/ip/tcp/tcp_transport_external.c.template]:
```suggestion
    // Place here your disconnection platform code
```

File: src/c/profile/transport/ip/udp/udp_transport_external.c.template [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+        const char* port)
+{
+    // Place here your initialization platform code
+    // Return true if success
+}
+
+bool uxr_close_udp_platform(
+        uxrUDPPlatform* platform)
+{
+    // Place here your close platform code
```
Comment (julibert) [inline src/c/profile/transport/ip/udp/udp_transport_external.c.template]:
```suggestion
    // Place here your closing platform code
```

File: src/c/profile/transport/ip/udp/udp_transport_external.c.template [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    // Return true if success
+}
+
+size_t uxr_write_udp_data_platform(
+        uxrUDPPlatform* platform,
+        const uint8_t* buf,
+        size_t len,
+        uint8_t* errcode)
+{
+    // Place here your write bytes platform code
```
Comment (julibert) [inline src/c/profile/transport/ip/udp/udp_transport_external.c.template]:
```suggestion
    // Place here your writing bytes platform code
```

File: src/c/profile/transport/ip/udp/udp_transport_external.c.template [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+size_t uxr_read_udp_data_platform(
+        uxrUDPPlatform* platform,
+        uint8_t* buf,
+        size_t len,
+        int timeout,
+        uint8_t* errcode)
+{
+    // Place here your read bytes platform code
```
Comment (julibert) [inline src/c/profile/transport/ip/udp/udp_transport_external.c.template]:
```suggestion
    // Place here your reading bytes platform code
```

File: src/c/profile/transport/serial/serial_transport_external.c.template [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+
+bool uxr_init_serial_platform(struct uxrSerialPlatform* platform, int fd, uint8_t remote_addr, uint8_t local_addr)
+{
+    // Place here your initialization platform code
+    // Return true if success
+}
+
+bool uxr_close_serial_platform(struct uxrSerialPlatform* platform)
+{
+    // Place here your close platform code
```
Comment (julibert) [inline src/c/profile/transport/serial/serial_transport_external.c.template]:
```suggestion
    // Place here your closing platform code
```

File: src/c/profile/transport/serial/serial_transport_external.c.template [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+    // Return true if success
+}
+
+size_t uxr_write_serial_data_platform(uxrSerialPlatform* platform, uint8_t* buf, size_t len, uint8_t* errcode)
+{
+    // Place here your write bytes platform code
```
Comment (julibert) [inline src/c/profile/transport/serial/serial_transport_external.c.template]:
```suggestion
    // Place here your writing bytes platform code
```

File: src/c/profile/transport/serial/serial_transport_external.c.template [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+    // Return number of wrote bytes
+}
+
+size_t uxr_read_serial_data_platform(uxrSerialPlatform* platform, uint8_t* buf, size_t len, int timeout, uint8_t* errcode)
+{
+    // Place here your read bytes platform code
```
Comment (julibert) [inline src/c/profile/transport/serial/serial_transport_external.c.template]:
```suggestion
    // Place here your reading bytes platform code
```

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Linux/84/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Windows/85/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Linux/88/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Windows/89/)

**Comment by pablogs9** [comment]
Rebased

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Linux/89/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Windows/90/)

**Pull request merged**